### PR TITLE
Add execution platform information to the action key.

### PR DESCRIPTION
**Pull request by katre**
Needed so that actions are properly invalidated and re-run when the platform's remote execution properties change.
 See 
Change-Id: I603d78efb0953b89682cc32226e246ed94af2e44

### CODE REVIEW

**Comment by katre** [comment]
@lberki I don't even know how to check that I updated all places that need this. Is there a centralized place to fix action key computation?

File: src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java [hunk 1]
```diff
@@ -179,6 +181,19 @@ public static Builder builder() {
...
 
+  /**
+   * Add this platform to the given fingerprint.
+   */
+  public void addTo(Fingerprint f) {
+    f.addString(label.toString());
+    f.addStrings(constraints.values().stream()
+      .map(c -> c.label().toString())
+      .collect(Collectors.toList()));
+    if (remoteExecutionProperties != null) {
```
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java]: This way, you can get the same strings added to the fingerprint if you remove remoteExecutionProperties but add a string to constraints.values(). Instead, use:

f.addString(label.toString())
f.addNullableString(remoteExecutionProperties);
f.addInt(constraints.size());
<add all the values>

Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java]: Also, why not add the keys of the map?

**Comment by lberki** [comment]
Sadly, I don't know of a centralized place :( The plan of record is to make everything a SpawnAction, but that doesn't help you right now.

**Comment by lberki** [comment]
Capturing our private chat: you can update `AbstractAction#getKey()` so that it adds the platform info to they key instead of using the result of `#computeKey()` directly. That way, you only need to change one place and can be sure that you got everything.

File: src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java (hunk 1 above)
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java]: The keys are actually based on the values, and only exist to make lookup efficient. The actual rule is only determines by the list of constraint values.

Updated the method.

**Comment by katre** [comment]
Unfortunately, I can't do anything with AbstractAction#getKey() because AbstractAction#computeKey returns a String, not a Fingerprint. Trying to change the way getKey/computeKey are implemented seems very invasive and I'd prefer to leave that alone, so leaving this as-is.

**Comment by lberki** [comment]
Why not put that String into a second Fingerprint then add a TODO to tell future adventurers that we have two Fingerprints just "accidentally"? Isn't optimal, but reasonably limited in scope.

**Comment by katre** [comment]
While technically correct, that feels really hacky to me.

**Comment by lberki** [comment]
Can't disagree with that, but which one do you prefer? Having a slightly hacky but workable solution with a clear path towards making it not actually hacky (by making `computeKey()` return a Fingerprint instead of a String) or repeating the same change over and over with (as you said) no guarantees that you repeated it enough times?

**Comment by katre** [comment]
Yeah. Expect a very large change tomorrow. Or maybe the next day....

**Comment by lberki** [comment]
You can also split it in two: the functional change with the hacky-but-compact solution and the huge change of making `computeKey()` return a Fingerprint (or take a Fingerprint as an arg). I'd much rather have huge no-op changes than a functional change hiding in a dusty corner of an otherwise no-op change. Then you also get to fix the issue faster :)

**Comment by katre** [comment]
I have an in progress change that uodates computeKey which I started last
night. Of course, there are about 50 call sites to modify, so finishing
that may take a while....

**Comment by katre** [comment]
I've rebased this onto 614fe0dfb9e6bed90c361e4b6bfff37c11a4775f and it's much easier to read and hopefully easier to prove correct.

File: src/main/java/com/google/devtools/build/lib/actions/AbstractAction.java [hunk 2]
```diff
@@ -314,6 +314,13 @@ public final synchronized String getKey(ActionKeyContext actionKeyContext) {
       try {
         Fingerprint fp = new Fingerprint();
         computeKey(actionKeyContext, fp);
+
+        // Always add the execution platform.
```
[the remark this answers is not in the document]
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/actions/AbstractAction.java]: Fixed.

File: src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java [hunk 3]
```diff
@@ -176,6 +178,15 @@ public static Builder builder() {
     return new Builder();
   }
 
+  /**
+   * Add this platform to the given fingerprint.
+   */
+  public void addTo(Fingerprint f) {
+    f.addString(label.toString());
+    f.addNullableString(remoteExecutionProperties);
+    f.addInt(constraints.size());
```
[the remark this answers is not in the document]
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java]: Fixed.

[the remark this answers is not in the document]
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java]: I'm somewhat out of the loop here -- is there a guarantee that `getExecutionInfo().constraints()` (and `.label()`) cannot influence the result of the action? At least naively, a subclass could just call that method and do different things based on that.

[the remark this answers is not in the document]
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java]: There's no "getExecutionInfo()" in this class.

The idea here is that if the constraints are changed, or if the remote execution properties are changed, the action probably needs to be re-run even if everything else (especially the generated command-line) is the same.

The constraints are specified in a rule as a list of constraint_value targets, then converted into a map in PlatformInfo (based on each constraint value's unique constraint setting) for ease of lookup.

There is the case where the constraint value is updated (possibly the constraint setting is changed), but not the platform, in which case re-analysis will update the Action, but here the key would be unchanged.

So, a solution: we just keep recursing downwards, as usual.

[the remark this answers is not in the document]
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java]: Erm, I misspoke *and* I think I don't understand something really trivial about this. Sorry for being dense. Let me try again.

Any AbstractAction subclass can call `this.getExecutionPlatform().constraints()` and make decisions based on the resulting map and the `ConstraintValueInfo` values in it. These are currently not added to the action key, thus making it possible for the action to change behavior without changing the key. Ways out:
 1. Enforce that any change in `PlatformInfo.constraints` causes `PlatformInfo.remoteExecutionProperties` to also change
 2. If `PlatformInfo.constraints()` cannot influence the action, remove the reference to it
 3. If it can, add `.constraints()` to the action key (this is the trivially correct course of action)

Do I understand correctly that your last paragraph means that the constraint value can change without the remote execution info changing? This would imply that (2) is the solution, I guess?

[the remark this answers is not in the document]
Comment (katre) [reply src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java]: When the execution platform changes (whether the constraints are changing, or the remote platform properties), the action _might_ change or it might not. There's no good way to tell (and in the case of the remote execution properties, only the remote execution system knows), so we need to always have a new action key.

This presents the problem that spurious changes to the execution platform will drown the cache, but in practice I don't expect execution platforms to change often. the motivating problem that I am trying to fix (and which I realize doesn't have an issue opened yet) is that a user updated the remote execution properties, and then Bazel used the cached result instead of telling the remote execution system to re-run the action, which would presumably have given different results.

The best answer here is 3) (and is what I am trying to implement). Ideally, 4) would also be true, but as you say that's not actually enforceable.

Opened #4756 to track this.

[the remark this answers is not in the document]
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java]: In that case, I shall approve this anticipating a followup change.
